hs-bindgen/src/HsBindgen/C/AST/Type.hs

```diff
@@ -38,7 +38,22 @@ data Type =
   | TypeEnum CName
   | TypeTypedef CName
   | TypePointer Type
-  | TypeConstArray Natural Type
+
+    -- | Arrays
+    --
+    -- Arrays normally have a known size, but not always:
+    --
+    -- * Arrays of unknown size are allowed as function arguments; such arrays
+    --   are interpreted as pointers.
+    -- * Arrays of unknown size may be declared for externs; this is considered
+    --   an incomplete type.
+    -- * Structs may contain an array of undefined size as their last field,
+    --   known as a "flexible array member" (FLAM).
+    --
+    -- We treat the FLAM case separately.
+    --
+    -- See <https://en.cppreference.com/w/c/language/array#Arrays_of_unknown_size>
+  | TypeArray (Maybe Natural) Type
```

Comment:
FWIW, introduce incomplete ararys in `Type` is very similar to having `void`there.

incomplete arrays cannot appear everywhere, and it somewhat depends how we'll treat them depending where we encounter them.

For that reason I'd kept `TypeConstArray` and added `TypeIncompleteArray` as a separate constructor, but that's more or less a style issue. (An argument is that we'll always want to match on the `Maybe` argument of `TypeArray` - const and incomplete arrays aren't treated anywhere the same).

Reply:
Ok, I'm fine with that, will change.